### PR TITLE
Support detection of native modules via NPM

### PR DESCRIPTION
**Pull request by sgtcoolguy**
**WORK IN PROGRESS**

**Description**:
This hooks support for detecting native Titanium modules installed via NPM.
The idea here is that we continue to support "legacy" modules (and auto-extract zip files of them) in the SDK and project. Additionally, we assume the first searchPath passed in is the project root, and we will look for modules under the project's `node_modules` folder. We only scan the top-level packages (or scoped packages), looking for a package.json. If the package has a package.json, has a `titanium` property, a `titanium.moduleid` property and `titanium.type === 'native-module'`, we assume the package to be a native module.

The expectation is that "new" native modules distributed this way are single-platform versions - and that the metadata typically found in a manifest file is found in the package.json.

Here's the example one I used:
```json
{
	"name": "@seadub/native-module",
	"version": "2.0.1",
	"description": "Pretend Native Module for Titanium SDK",
	"author": "Chris Williams",
	"license": "Apache",
	"copyright": "Copyright (c) 2015 by Appcelerator Inc.",
	"titanium": {
		"type": "native-module",
		"guid": "bba89061-0fdb-4ff1-95a8-02876f5601f9",
		"name": "Native Module",
		"moduleid": "native-module",
		"platform": "ios",
		"apiversion": 2,
		"architectures": [ "armv7", "arm64", "i386", "x86_64" ],
		"minsdk": "5.0.0"
	}
}
```

**Additional notes**
- This raises the min node to 8.0
- I attempted to refactor to es2017 syntax in timodule, so I'm making use of async/await/promises rather than the `async` lib there.
- This modifies behavior slightly to make any platform name references of `'iphone'` or `'ipad'` become `'ios'` so that the returned object **always** uses `'ios'` as the platform name/key.
- The contents of the package are assumed to be the same as what you'd find in a module zipfile/dir with the additional package.json file. I didn't try to tackle having the package.json point to the JAR files/static libraries/frameworks/etc.

**Ok, what's left to explore?**
- Do we support globally installed npm packages/modules?
  - I have this working locally, but if we need to sniff the npm global install dir properly, it requires launching a process to run `npm root -g`, which ends up slowing detection way down (for me it took ~275-300ms to run this command, which ended up being ~99.5% of the total execution time per test). We could cheat and assume the typical install dirs, but that'd break us for anyone changing their npm prefix config value.
- How do we handle composite/multi-platform modules?
  - I assume we could distribute those as an npm package whose dependencies are the individual per-platform packages, and have the composite one use a different `titanium.type` value in its package.json (say `'composite-native-module'`).
  - The question is whether or not this is even needed. it's been a hack to do these so far.
- How does this work if we move from tiapp.xml to package.json for Titanium projects?
  - The individual native modules would become dependencies in the project's package.json, and it should *just work*. NPM would then effectively handle semver resolution for us, and users would have to be explicit with their module dependencies.
- Module dependencies
  - we have some existing support now in our CLI. How would that work here? Do we define other native module packages as dependencies? (if so, then they may not be top-level dirs so we wouldn't pick them up here) Do they become peerDependencies?
- What about the existing manifest file?
  - Can we also support a native module that has a bare minimum package.json (i.e. no `titanium` section) but also has a manifest file? Is it worth it?

### CODE REVIEW

**Comment by janvennemann** [comment]
Nice one! 

> Do we support globally installed npm packages/modules?

I'd say no. Locally installed packages should be totally fine. Installing a package with NPM is super easy so now that i don't have to extract zip files and copy stuff around anymore i don't see much of a benefit in globally installed module packages.

> Module dependencies

Defining them as peerDependencies sound like the right choice to me.

> What about the existing manifest file?

Make a hard cut here i'd say. If it is distributed via NPM it needs a package.json with a valid `titanium` section.

> How do we handle composite/multi-platform modules?

So this is an important topic i think! I would really love to eventually see one package that contains both platforms. However, i'm not really a fan of a composite package and a package for each platform.

While working on a few modules lately i felt that is was a pain in the a** to manage the different platforms. iOS is at version 2.X, while Android is already at 4.X for example. The APIs are different, i have to manage and release for each platform independently and so on. Now with a composite package i would have to manager three different packages.

This totally works against our whole idea if cross-platform compatibility and is not really user friendly IMHO. I think if we move away from separate module packages per platform we can push users into the direction of keeping it cross-platform.

For example, take this simple package structure:
```
module-name
├── android
│   ├── lib
│   ├── libs
│   ├── module-name.jar
│   └── ...
├── ios
│   ├── lib
│   ├── libmodule-name.a
│   ├── module.xconfig
│   └── ...
├── 
├── package.json
├── README.json
└── ...
```

Basically just copy the Android and iOS dist into a subfolder and then read it from there if its an NPM package. The only platform specific things from the manifest could be grouped under a `platforms` key in `package.json`:

```
{
  "name": "my-module",
  "version": "1.0.0",
  "titanium": {
    "platforms": {
      "android": {
        "apiversion": 4
        "architectures": [ "arm64-v8a", "armeabi-v7a", "x86" ]
      },
      "ios": {
        "apiversion": 2
        "architectures": [ "armv7", "arm64", "i386", "x86_64" ]
      }
    }
  }
}
```

**Comment by brentonhouse** [comment]
I would not recommend using peerDependencies as they were deprecated with 3.x.   https://docs.npmjs.com/files/package.json#peerdependencies

I would instead make them regular dependencies.   devDependencies will be things like grunt, eslint, etc that are needed for development but not compiling.

**Comment by brentonhouse** [comment]
One of the issues with including both Android and iOS in the same package is versioning.  Are both platforms always going to have the same version number?  i.e.  if Android module introduces a breaking change, do we increment the major version for the package (which includes the iOS module)?

**Comment by brentonhouse** [comment]
Separate note, but because this could be used to install non-native modules and widgets in the future, we probably want to be explicit about the fact that this is a native module in the package.json titanium section.   If the `native` flag is not there, I would copy that package (and dependencies) to `node_modules` in the `Resources` directory tree.    

I created a quick package here:  https://github.com/brentonhouse/titanium-module-copier
as an example of copying packages but would need to modify it to ignore any packages that are flagged with `native`.

**Comment by cb1kenobi** [comment]
Thanks for tackling this problem @sgtcoolguy!

> Do we support globally installed npm packages/modules? 

I do not think we should support modules installed in the global npm directory. However, I do think we should support modules in global Titanium modules directory (e.g. `~/Library/Application Support/Titanium/modules`) and possibly some other global AMPLIFY CLI Titanium Module package directory.

The AMPLIFY CLI Package Manager currently supports downloading modules from GitHub via the Marketplace (AppDirect). The Marketplace is not going away and I doubt the Titanium Modules in the store will go away. I can see people distributing their PAID modules via the Marketplace and wanting to use this new module project format. Module authors will want the benefits of the `package.json`, semver, and dependency management.

As far as the AMPLIFY CLI PM is concerned, it can simply check if the extracted files contains a `package.json` and run `npm i`. I don't think we need a new package type in the Registry Server.

> How do we handle composite/multi-platform modules? 

I prefer every platform implementation in the same package. API compatibility/parity is very important If the iOS code causes a major rev, then Android revs too. I don't see a problem with that.

In my native C++ Node.js addons, the main entry point is a .js file, so it seems natural that Titanium modules would have a `main` in the `package.json` that points to an `index.js` or something that then exposes the native modules.

If Titanium module's public API is .js, then we make it MUCH easier to promote cross-platform interfaces that hide platform differences.

One thing we need to be wary of is encryption. We don't have a good way today of encryption .js files that are distributed with a native module. Yes we have encrypted commonjs modules, but I have no idea if that still works today or if anybody actually knows it exists. I think when we redesign the encryption stuff, we take into consideration .js files in modules. I suspect we can do the same process as apps where each file is encrypted and no encrypted code is actually embedded in the binary.

I personally do NOT like putting the metadata in the `package.json`, but I don't care enough to argue about this.

> How does this work if we move from tiapp.xml to package.json for Titanium projects?

I strongly oppose moving from the `tiapp.xml` to a `package.json`. We need both. `package.json` would be for global project settings (name, version, etc) and dependencies. But I think we also need a separate project file to track the more nitty gritty project settings: encryption, properties, which modules are enabled, provisioning profile uuids, etc. Plus we need a 3rd file which is a user-specific, non-version controlled config file that overrides the project file. I would prefer the format of the project files be .xml or .js, not .json. The ability to comment out lines is essential.

> Module dependencies

If modules declare a `main` in the `package.json`, it makes the public API .js and that's very easy for one module to consume another. I think.

> What about the existing manifest file?

We will likely have to support the manifest file for many years. Load the one off disk, then merge the one from the `package.json` on top of it. Done.

**Comment by sgtcoolguy** [comment]
> I would not recommend using peerDependencies as they were deprecated with 3.x. https://docs.npmjs.com/files/package.json#peerdependencies
> 

Just a note that peerDependencies are not deprecated, just that they aren't automatically installed for the end-user, instead if not found in the project's list of dependencies, it'll warn the user on `npm install` that it's missing.

Using peerDependencies means I wouldn't have to dive deeper into the directory structure to detect modules - which is not a big deal to fix. But I do think using peerDependencies for something like `ti.playservices` is the right thing to do since it should be "shared" by say `ti.map` and other modules.

**Comment by sgtcoolguy** [comment]
> Separate note, but because this could be used to install non-native modules and widgets in the future, we probably want to be explicit about the fact that this is a native module in the package.json titanium section. If the `native` flag is not there, I would copy that package (and dependencies) to `node_modules` in the `Resources` directory tree.
> 
> I created a quick package here: https://github.com/brentonhouse/titanium-module-copier
> as an example of copying packages but would need to modify it to ignore any packages that are flagged with `native`.

Agreed, which is why I have the `titanium.type` property and look for a value of `'native-module'`. I assume if we spec this out further and think of widgets/project templates/etc we'd need to flag them as different specialized packages.

**Comment by sgtcoolguy** [comment]
> Thanks for tackling this problem @sgtcoolguy!
> 
> > Do we support globally installed npm packages/modules?
> 
> I do not think we should support modules installed in the global npm directory. However, I do think we should support modules in global Titanium modules directory (e.g. `~/Library/Application Support/Titanium/modules`) and possibly some other global AMPLIFY CLI Titanium Module package directory.

I think we're in agreement here. This change still retains our existing behavior of loading modules from the project and SDK directories in their zip file or extracted form (with no `package.json`, but with `manifest` files) - and the list of directories is possible to expand via cli config.

> The AMPLIFY CLI Package Manager currently supports downloading modules from GitHub via the Marketplace (AppDirect). The Marketplace is not going away and I doubt the Titanium Modules in the store will go away. I can see people distributing their PAID modules via the Marketplace and wanting to use this new module project format. Module authors will want the benefits of the `package.json`, semver, and dependency management.

Again, we agree - I think for "non-free" (paid, licensed, private) modules they'll need to use our marketplace to handle that. But yeah, this may impact the AMPLIFY CLI manager needing to handle modules with package.json files.
This was an attempt to handle the much more common case of "free" modules to make consumption/distribution much easier with minimal changes.

**Comment by sgtcoolguy** [comment]
Ok, so far the takeaways seem to be:
- globally installed npm-based native modules seem OK to not support (yay!)
- We/I should come up with a better proposal for cross-platform native modules via npm than my "composite" idea. @janvennemann 's suggestion closely mirrors what I was originally playing around with locally when I was going to use hyper loop as the model, but dropped to focus on a simpler case first.
- Will the CLI/NPM handle if we have two different versions of the same module in the node_modules tree? (What if there's a conflict between two modules requiring different versions of the same native module dependency)
- I think @cb1kenobi 's idea of merging the manifest with the package.json makes sense and I'll take a look at that. My only concern is if the two "disagree" on values/they are out of sync.

**Comment by sgtcoolguy** [comment]
ok, so I've pushed a couple changes:
- If a manifest file exists we will merge the values with those found in package.json. In this case`package.json` will "win" over `manifest` if they both contain values. I do not throw Errors if they each have values but they don't match. 
- Support for native modules with multiple platforms. It is assumed that the module will have a sub-directory for each platform.

Note that combined this makes the code more forgiving and offers a much simpler bare-bones `package.json` now. If you had an existing module to migrate you'd need to generate a package.json with the name/version for npm, and the `titanium.type`(with value `'native-module'`) and `titanium.platform` properties. I now support platform values of a single string, an array of strings, or an object whose keys are the platform names and value is an object holding the manifest metadata. If there's a single platform, we can support either a sub-dir holding the contents, or the top-level of the module.

So at bare minimum you could do:
```json
{
  "name": "example",
  "version": "1.0.0",
  "titanium": {
    "type": "native-module",
    "platform: "ios"
  }
}
```
with the contents you'd typically find under the 'ios' subdirectory, including the `manifest` file and we'll use that file's contents to fill in the id/minsdk/architectures/etc.

As for @cb1kenobi 's suggestion of pointing the top-level `main` at a JS file that exposes the API, I have not attempted to tackle that. That seems more of an SDK & CLI change to treat native modules more like we do CommonJS modules.

**Comment by janvennemann** [comment]
> One of the issues with including both Android and iOS in the same package is versioning. Are both platforms always going to have the same version number? i.e. if Android module introduces a breaking change, do we increment the major version for the package (which includes the iOS module)?

The goal should be to not introduce breaking changes just for a single platform. This goes against cross-platform compatibility. If the module API needs breaking changes, they should be applied for both platforms so they stay the same.

**Comment by sgtcoolguy** [comment]
I've manually merged this over and cut a new node-app 0.3.0 release.
